### ISSUE TITLE
use a different service mesh

### ISSUE THREAD

**Issue by rverma-nikiai** [kubeflow/kubeflow#3396]
We are using aws app-mesh currently and we are very willing to use kubeflow. The default installation through kfctl.sh installs duplication alb-ingress-controller (we already have one deployed) and complete istio-system. Will it be possible to build a deployment guide which allows using an alternate service-mesh.

I would be happy to add code contribution and testing for the same.

**Comment by kkasravi** [comment]
@rverma-nikiai we're deploying istio as a kustomize overlay (see https://github.com/kubeflow/manifests/pull/102) which would allow a different service mesh to be configured (as an alternative overlay). 

Would this be part of the aws install (where you install aws app-mesh)?

**Comment by rverma-nikiai** [comment]
I install aws app-mesh after the eks cluster being set up. It's an optional feature and requires to install something like appmesh-system to install using kubectl (very similar to istio-system). Follows the conventions of virtual services also.

**Comment by kkasravi** [comment]
ok, once https://github.com/kubeflow/manifests/pull/102 and https://github.com/kubeflow/kubeflow/pull/3375 are merged it should be straightforward to copy how istio overlays are implemented and do something similar with aws app-mesh. 
@jlewi @kunmingg any comments on entertaining the idea of an aws service-mesh that could be added in this way?

**Comment by rverma-nikiai** [comment]
Will this allow to install into existing istio setup. If yes I can try to create a document for setup in aws app mesh. Also while we are on it, it would be nice to be able to use external MySQL.

**Comment by kkasravi** [comment]
do you mean replace existing istio setup?

there is a mysql component installed by pipeline and another by argo i believe. It would just be a matter of adding the component to the platform. I can provide more info once the above PRs are merged

**Comment by rverma-nikiai** [comment]
Yes, while doing a poc with the istio setup I installed istio 1.1.7 minimal with helm. When I tried deploying kubeflow it replaced the istio setup(1.1.6) and installed more components like grafana, prometheus, kiali, etc. which I doubt were required for kubeflow. I would also encourage if we are using a limited part of istio (I believe only the pilot is required) new vanilla setup should install istio-minimal, also if istio is completely missing and if istio is already installed shouldn't do anything.

This should be done declaratively like the flag `useIstio == true`,  at the same time it doesn't give control that what components we can install. The rationale being while the time I figure out native app-mesh support for kubeflow, I still want kubeflow to run with bare minimum istio(which is required for kubeflow only). 

It would be really nice to support this as a helm chart where all other requirements including istio work as a dependency. Not only this is widely accepted, it also saves you the pain of maintaining dependent components. E.g we are also using Argo, we prefer to disable MySQL etc. It gives freedom of overriding the dependent chart values as well, e.g. if someone wants to try the latest tensorflow-serving or jupyter-notebook, at the same time you can create some configmap resources to wire default config required for kubeflow. 

Also, I am wondering what's wrong with a simple ingress resource with path-based ingress. Why we even need to rely on istio?

I strongly believe utilizing helm as distribution will definitely improve adoption among the community which recently adopted kubernetes.
This is more of my opinionated view, I would love to understand your perspective as well. 
Thanks

**Comment by kkasravi** [comment]
> Yes, while doing a poc with the istio setup I installed istio 1.1.7 minimal with helm. When I tried deploying kubeflow it replaced the istio setup(1.1.6) and installed more components like grafana, prometheus, kiali, etc. which I doubt were required for kubeflow. I would also encourage if we are using a limited part of istio (I believe only the pilot is required) new vanilla setup should install istio-minimal, also if istio is completely missing and if istio is already installed shouldn't do anything.

This perspective is also echoed by @jlewi - where, for example, if you're deploying kubeflow to an existing cluster (on-prem use case) there should be an option to install istio (or not if already there)

> It would be really nice to support this as a helm chart where all other requirements including istio work as a dependency. Not only this is widely accepted, it also saves you the pain of maintaining dependent components. E.g we are also using Argo, we prefer to disable MySQL etc. It gives freedom of overriding the dependent chart values as well, e.g. if someone wants to try the latest tensorflow-serving or jupyter-notebook, at the same time you can create some configmap resources to wire default config required for kubeflow.

This [design doc](https://docs.google.com/document/d/1SVm-xffIdWiMYkl_nQO5C6HHK46NWkfDGPDUL5ukeME/edit#heading=h.frwscropjuwb) covers the rationale for using kustomize vs helm. That said, kustomize does provide plugin support for helm where a helm chart and its related vars could be rendered along side other kustomize resources.

A vendor | platform specific deployment can be done by editing the app.yaml after `kfctl init`. Per your use-case components could be added/removed and parameters changed. Alternatively, the app.yaml file is generated from a 'template' file under kubeflow/bootstrap/config, which is assembled by kustomize leveraging overlays. This is where platform yamls will live. Some specific details are still in-flux @jlewi can cover the design guidance he's providing me, but the idea is to allow anyone to define a deployment with their customizations.

> Also, I am wondering what's wrong with a simple ingress resource with path-based ingress. Why we even need to rely on istio? 

 @jlewi | @kunmingg (^)

**Comment by rverma-nikiai** [comment]
@kkasravi Makes total sense for using kustomize. Is there a development branch where I can play with this?

**Comment by kkasravi** [comment]
@rverma-nikiai 
you could fork and create a branch on your fork with upstream set to kubeflow.

There are [developer guidelines](https://github.com/kubeflow/kubeflow/blob/master/bootstrap/developer_guide.md) for building kfctl

**Comment by jlewi** [comment]
@rverma-nikiai Does that answer your question?

**Comment by jlewi** [comment]
Some of the information in this thread is outdated.

I thought we had a different issue/google thread with more discussion about ISTIO and alternatives but I can't find it (so maybe it was slack).

TL;DR if you want to use a different service mesh technology you'd want to treat ISTIO resources as abstract resources. The primary resources that Kubeflow depends on are VirtualService resources and we primarily use these to dynamically register reverse proxy routes.

So if you had a different networking stack you wanted to use one approach would be to

1. Only install the Virtual Service CRD (just the definition not the controller)
1. Create a custom controller for VirtualService that setup the appropriate reverse proxy route using your networking stack.
